fabric-input-events-v0/src/main/java/net/fabricmc/fabric/mixin/event/input/client/InputUtilMixin.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.mixin.event.input.client;
+
+import java.util.Map;
+
+import org.lwjgl.glfw.GLFW;
+import org.lwjgl.glfw.GLFWCharModsCallbackI;
+import org.lwjgl.glfw.GLFWKeyCallbackI;
+import org.lwjgl.glfw.GLFWCursorPosCallbackI;
+import org.lwjgl.glfw.GLFWDropCallback;
+import org.lwjgl.glfw.GLFWMouseButtonCallbackI;
+import org.lwjgl.glfw.GLFWScrollCallbackI;
+import org.lwjgl.glfw.GLFWDropCallbackI;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.ModifyVariable;
+
+import net.fabricmc.fabric.api.client.event.input.CharEvent;
+import net.fabricmc.fabric.api.client.event.input.ClientInputEvents;
+import net.fabricmc.fabric.api.client.event.input.KeyEvent;
+import net.fabricmc.fabric.api.client.event.input.KeybindEvent;
+import net.fabricmc.fabric.api.client.event.input.MouseButtonEvent;
+import net.fabricmc.fabric.api.client.event.input.MouseMoveEvent;
+import net.fabricmc.fabric.api.client.event.input.MouseScrollEvent;
+import net.fabricmc.fabric.impl.client.FabricKeyboardImpl;
+import net.minecraft.client.options.KeyBinding;
+import net.minecraft.client.util.InputUtil;
+
+@Mixin(InputUtil.class)
+public class InputUtilMixin {
+	@ModifyVariable(method="setKeyboardCallbacks(JLorg/lwjgl/glfw/GLFWKeyCallbackI;Lorg/lwjgl/glfw/GLFWCharModsCallbackI;)V", at=@At("HEAD"), index=2)
+	private static GLFWKeyCallbackI fabric_changeKeyCb(GLFWKeyCallbackI keyCb) {
+		return (long window, int code, int scancode, int action, int mods) -> {
```

Comment:
This seems to replace the entire key callback.

I would actually inject into the synthetic method that the lambda is in vanilla so we don't need to duplicate vanilla code as you only seem to inject in like 3 spots in the entire method

Reply:
I wasn't sure how that would work with other mods trying to do the same thing. This way we can be sure events get emitted before anyone else can screw things up. Also, the callback is still called, of course.